pkg/ccl/importccl/read_import_pgdump.go

```diff
@@ -567,22 +571,44 @@ func (m *pgDumpReader) readFile(
 			if ok && conv == nil {
 				return errors.Errorf("missing schema info for requested table %q", name)
 			}
+			expectedColLen := len(i.Columns)
+			if expectedColLen == 0 {
+				// Case where the targeted columns are not specified in the PGDUMP file, but in
+				// the command "IMPORT INTO table (targetCols) PGDUMP DATA (filename)"
+				expectedColLen = len(conv.VisibleCols)
+			}
 			values, ok := i.Rows.Select.(*tree.ValuesClause)
 			if !ok {
 				return errors.Errorf("unsupported: %s", i.Rows.Select)
 			}
 			inserts++
 			startingCount := count
+			colLookup := make(map[string]int, len(conv.VisibleCols))
+			for j, col := range conv.VisibleCols {
+				colLookup[col.Name] = j
+			}
 			for _, tuple := range values.Rows {
 				count++
 				if count <= resumePos {
 					continue
 				}
-				if expected, got := len(conv.VisibleCols), len(tuple); expected != got {
-					return errors.Errorf("expected %d values, got %d: %v", expected, got, tuple)
+				if got := len(tuple); expectedColLen != got {
+					return errors.Errorf("expected %d values, got %d: %v", expectedColLen, got, tuple)
 				}
-				for i, expr := range tuple {
-					typed, err := expr.TypeCheck(ctx, &semaCtx, conv.VisibleColTypes[i])
+				conv.IsTargetCol = make(map[int]struct{})
```

Comment:
We need a way to stop doing multiple map creations per INSERT row. This has now regressed and will recreate this IsTargetCol map per INSERT row, whether or not columns are specified, which is worse than what we had before.

Reply:
I've moved the `conv.IsTargetCol` creation (more like an update) to an outer loop so that it will be done for every `INSERT` statement instead of each _row_ of the `INSERT` statement. Unfortunately, this is the best possible optimization in terms of the frequency of the map creation as the target columns (and possibly the order) of each insert statement might be different. 

I've also made `targetColMapInd` slice for each `INSERT` statement so that each tuple of each row can know which data each tuple corresponds to.